File: Daily/380.insert-delete-get-random-o-1.js
/*
 * @lc app=leetcode id=380 lang=javascript
 *
 * [380] Insert Delete GetRandom O(1)
 */

// @lc code=start

class RandomizedSet {
    /** @type {Map<number, number>}*/
    #numberToIndex;
    /** @type {number[]} */
    #numbers;

    constructor() {
        this.#numberToIndex = new Map();
        this.#numbers = [];
    }

    /**
     * @param {number} val
     * @return {boolean}
     */
    insert(val) {
        if (this.#numberToIndex.has(val)) return false;

        this.#numbers.push(val);
        this.#numberToIndex.set(val, this.#numbers.length - 1);
        return true;
    }
    /**
     * @param {number} val
     * @return {boolean}
     */
    remove(val) {
        if (!this.#numberToIndex.has(val)) return false;

        const lastNumber = this.#numbers[this.#numbers.length - 1];

        if (lastNumber !== val) {
            // swap the removed number to be the last one
            const lastNumberIndex = this.#numbers.length - 1;
            const removedNumberIndex = this.#numberToIndex.get(val);

            [
                this.#numbers[removedNumberIndex],
                this.#numbers[lastNumberIndex],
            ] = [
                this.#numbers[lastNumberIndex],
                this.#numbers[removedNumberIndex],
            ];

            this.#numberToIndex.set(lastNumber, removedNumberIndex);
        }

        this.#numbers.pop();
        this.#numberToIndex.delete(val);
        return true;
    }
    /**
     * @return {number}
     */
    getRandom() {
        const length = this.#numbers.length;
        const randomIndex = this.#generateRandomInteger(0, length);
        return this.#numbers[randomIndex];
    }

    /**
     * Generate random integer in [begin, end)
     * @param {number} begin
     * @param {number} end
     * @returns {number}
     */
    #generateRandomInteger(begin, end) {
        const diff = end - begin;
        return begin + Math.floor(Math.random() * diff);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */
// @lc code=end

File: prev-2024/380.insert-delete-get-random-o-1.cpp
/*
 * @lc app=leetcode id=380 lang=cpp
 *
 * [380] Insert Delete GetRandom O(1)
 */

#include <cstdlib>
#include <ctime>
#include <map>
#include <vector>

// @lc code=start
using std::map;
using std::vector;

class RandomizedSet {
 private:
  vector<int> elements;
  map<int, int> elementToIndex;

  int getRandomNumber(int start, int end) {
    const int diff = end - start;
    return std::rand() % (diff + 1) + start;
  }

 public:
  bool insert(int val) {
    if (!this->elementToIndex.contains(val)) {
      this->elements.push_back(val);
      this->elementToIndex[val] = this->elements.size() - 1;
      return true;
    }
    return false;
  }

  bool remove(int val) {
    if (!this->elementToIndex.contains(val)) {
      return false;
    }

    const int lastElementIndex = this->elements.size() - 1;
    const int lastElement = this->elements.at(lastElementIndex);
    const int valIndex = this->elementToIndex[val];

    if (valIndex != lastElement) {
      this->elements[lastElementIndex] = val;
      this->elements[valIndex] = lastElement;
      this->elementToIndex[val] = lastElementIndex;
      this->elementToIndex[lastElement] = valIndex;
    }

    this->elements.pop_back();
    this->elementToIndex.erase(val);
    return true;
  }

  int getRandom() {
    const int randomIndex = this->getRandomNumber(0, this->elements.size() - 1);
    return this->elements[randomIndex];
  }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */
// @lc code=end

int main() {
  RandomizedSet* rset = new RandomizedSet();
  rset->insert(1);
  rset->remove(2);
  rset->insert(2);
  rset->getRandom();
  rset->remove(1);
  rset->insert(2);
  rset->getRandom();
}
